src/dbnode/storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -736,41 +1030,139 @@ func (s *fileSystemSource) bootstrapDataRunResultFromAvailability(
-	err     error
+type bootstrapFromIndexPersistedBlocksResult struct {
+	fulfilled result.ShardTimeRanges
+	result    *runResult
 }
 
-func newShardReaders(shard uint32, tr xtime.Ranges, readers []fs.DataFileSetReader) shardReaders {
-	return shardReaders{
-		shard:   shard,
-		tr:      tr,
-		readers: readers,
+func (s *fileSystemSource) bootstrapFromIndexPersistedBlocks(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+) (bootstrapFromIndexPersistedBlocksResult, error) {
+	res := bootstrapFromIndexPersistedBlocksResult{
+		fulfilled: result.ShardTimeRanges{},
+	}
+
+	indexBlockSize := ns.Options().IndexOptions().BlockSize()
+	infoFiles := fs.ReadIndexInfoFiles(s.fsopts.FilePathPrefix(), ns.ID(),
+		s.fsopts.InfoReaderBufferSize())
+
+	for _, infoFile := range infoFiles {
+		if infoFile.Err.Error() != nil {
+			s.log.WithFields(
+				xlog.NewField("namespace", ns.ID().String()),
+				xlog.NewField("error", infoFile.Err.Error()),
+				xlog.NewField("shardsTimeRanges", shardsTimeRanges.String()),
+				xlog.NewField("filepath", infoFile.Err.Filepath()),
+			).Error("unable to read index info file")
+			continue
+		}
+
+		info := infoFile.Info
+		indexBlockStart := xtime.UnixNano(info.BlockStart).ToTime()
+		indexBlockRange := xtime.Range{
+			Start: indexBlockStart,
+			End:   indexBlockStart.Add(indexBlockSize),
+		}
+		willFulfill := result.ShardTimeRanges{}
+		for _, shard := range info.Shards {
+			tr, ok := shardsTimeRanges[shard]
+			if !ok {
+				// No ranges match for this shard
+				continue
+			}
+
+			iter := tr.Iter()
+			for iter.Next() {
+				curr := iter.Value()
+				intersection, intersects := curr.Intersect(indexBlockRange)
+				if !intersects {
+					continue
+				}
+				willFulfill[shard] = willFulfill[shard].AddRange(intersection)
+			}
+		}
+
+		if willFulfill.IsEmpty() {
+			// No matching shard/time ranges with this block
+			continue
+		}
+
+		segments, err := fs.ReadIndexSegments(fs.ReadIndexSegmentsOptions{
+			ReaderOptions: fs.IndexReaderOpenOptions{
+				Identifier:  infoFile.ID,
+				FileSetType: persist.FileSetFlushType,
+			},
+			FilesystemOptions: s.fsopts,
+		})
+		if err != nil {
+			s.log.WithFields(
+				xlog.NewField("namespace", ns.ID().String()),
+				xlog.NewField("error", err.Error()),
+				xlog.NewField("blockStart", indexBlockStart.String()),
+				xlog.NewField("volumeIndex", infoFile.ID.VolumeIndex),
+			).Error("unable to read segments from index fileset")
+			continue
+		}
+
+		// Track success
+		s.metrics.persistedIndexBlocksRead.Inc(1)
+
+		// Record result
+		if res.result == nil {
+			res.result = newRunResult()
+		}
+		segmentsFulfilled := willFulfill
+		indexBlock := result.NewIndexBlock(indexBlockStart, segments,
+			segmentsFulfilled)
+		res.result.index.Add(indexBlock, nil)
+		res.fulfilled.AddRanges(segmentsFulfilled)
 	}
+
+	return res, nil
+}
+
+type timeWindowReaders struct {
+	ranges  result.ShardTimeRanges
+	readers map[shardID]shardReaders
+}
+
+type shardID uint32
```

Comment:
+1 for the alias, we should make this a first class alias in `storage` sometime down the road. i'll open a ticket to track

Reply:
Yeah true, we should make it more prevalent.  Sounds good 👍